### ISSUE TITLE
Restrict non-boolean properties with is prefix

### ISSUE THREAD

**Issue by imanushin** [detekt/detekt#2701]
## Expected Behavior of the rule
Deny property, which starts with ```is``` (e.g. name is like ```isDefault```), however the return value isn't ```Boolean```.

E.g. the right property sample: ```isValid: Boolean```. The invalid property sample: ```isValid: String```.

## Context
According to Java (not Kotlin) rules, only boolean properties should start with ```is```. Kotlin has the following behavior: for property ```isXXX```, the getter will be ```isXXX```, not ```getIsXXX```.

For example, for ```class Temp(val isValid: String)``` resulting decompiled bytecode ([please check via this solution](https://stackoverflow.com/a/40762755/185498)) will be:
```java
public final class Temp {
   @NotNull
   private final String isValid;

   @NotNull
   public final String isValid() {
      return this.isValid;
   }

   public Temp(@NotNull String isValid) {
      Intrinsics.checkParameterIsNotNull(isValid, "isValid");
      super();
      this.isValid = isValid;
   }
}
```

From Java Language Specification, ```is``` prefix is allowed for ```Boolean``` properties only, please check chapter 8.3.2 at [the document](https://download.oracle.com/otn-pub/jcp/7224-javabeans-1.01-fr-spec-oth-JSpec/beans.101.pdf?AuthParam=1589736840_2ff06f9f65563391bd83768de1b0ab41).

**Finally: the method ```isValid```, which returns ```String```, isn't correctly named.**

## Where it can be
Here is just example, this isn't generic behavior.

Let's imagine, that you want to send json into Oracle and parse it there (to achieve bulk operations benefits).

On the interoperation layer, some classes could be used just for serialization/deserialization from Json input/output. So such values can be allowed there (just for interoperation idea). For example, [Oracle doesn't support Boolean datatype](https://www.oreilly.com/library/view/oracle-plsql-programming/0596003811/ch12.html). So if you want to send json into Oracle and deserialize it there, you should use ```NUMBER(1)``` or ```CHAR(1)``` (e.g. you have to send 0/1 or Y/N to Oracle).

For example, for Kotlin dto ```class SomeValue(val isDefault: Boolean, val type: String)```, you have to create special dto for Oracle (you can't use Boolean). So, Oracle-friendly dto will be ```class SomeValue(val isDefault: Int, val type: String)``` (```isDefault``` can be only 0 or 1).

And this class is wrong, because:
1. ```val isDefault: Int``` isn't correct property from Java prospective
2. Some libraries aren't ready for this unusual naming, please see https://github.com/FasterXML/jackson-module-kotlin/issues/337 as an example.

**Comment by schalkms** [comment]
That's very interesting. This issue seems to be specific to JavaBeans connected to an Oracle database (PLSQL).

**Comment by imanushin** [comment]
@schalkms , no, this isn't related to Oracle. The main idea is "avoid invalid naming". I just show example, when ```isDefault: Int``` could be quite logical (however it still wrong and it is worth renaming this property).

**Comment by schalkms** [comment]
I see. I think it would nicely integrate into the `naming` ruleset. Can you again lend us a helping hand here @imanushin ?

**Comment by imanushin** [comment]
Will try to integrate, thank you !

**Comment by imanushin** [comment]
Just for history - I'm trying to get exact type from property. And there are some difficulties there. However, looks like, there is solution from https://discuss.kotlinlang.org/t/retrieve-kotlin-property-type-from-kotlin-psi-api/17874

**Comment by schalkms** [comment]
@imanushin that's a cool finding. Please stay tuned.

**Comment by arturbosch** [comment]
@imanushin you may want to use `property.getKotlinTypeForComparison(context)` for your case.
Make sure that the rule tests if `context != BindingContext.empy`.
This rule is a potential candidate to work with and without type resolution.

Code:
```kotlin
val bool = true
val bool2: Boolean = true
```

Psi:
```kotlin
    override fun visitProperty(property: KtProperty) {
        println(property.name)
        println(property.typeReference?.text)
        println(property.getKotlinTypeForComparison(context))
    }
``` 

Output:
```
bool
null
Boolean
bool2
Boolean
Boolean
```

**Comment by imanushin** [comment]
@arturbosch , wow, that's great, thank you! Will use.
